python/revng/pypeline/pipeline.py

```diff
@@ -386,20 +389,38 @@ def schedule(
             scheduled_task.incoming = scheduled_task.incoming.minimize()
             scheduled_task.outgoing = scheduled_task.outgoing.minimize()
 
-            # Figure out if a task will actually produce outputs
+            # Figure out if a task will actually produce outputs and which
+            # containers are read and written by it
             written_out = Requests()
+            written_containers: set[ContainerDeclaration] = set()
+            read_containers: set[ContainerDeclaration] = set()
             for argument in scheduled_task.node.arguments_with_access:
-                if argument.access != TaskArgumentAccess.READ:
-                    container_decl = argument.to_container_decl()
-                    if container_decl in scheduled_task.outgoing:
-                        written_out[container_decl] = scheduled_task.outgoing[container_decl]
+                container_decl = argument.to_container_decl()
+                access = argument.access
+                if access != TaskArgumentAccess.READ and container_decl in scheduled_task.outgoing:
+                    written_out[container_decl] = scheduled_task.outgoing[container_decl]
+
+                if access == TaskArgumentAccess.WRITE:
+                    written_containers.add(container_decl)
+                elif access & TaskArgumentAccess.READ:
+                    read_containers.add(container_decl)
```

Comment:
Suspicious piece of code.

1. What about `READ_WRITE`?
2. Why do you use `&` in the second check?

If correct, add comments.

Reply:
Both are there because `READ_WRITE` is handled
The first `if` basically determines if the container is "truncated", which allows to add another pipe as expiring
The second `if` basically checks "is the access READ or READ_WRITE", if so and it's the "last to read" then it marks that container as expiring